### PR TITLE
Improvements for Papers100m single gpu and single node multi gpu examples (Cugraph, GATConv, better default hyperparams, eval on all ranks)

### PR DESCRIPTION
**Pull request by puririshi98**
Introduces improvements to NVIDIA GPU examples:
- Adds support for cuGraph data loading
- Adds `GAT` in addition to `GCN`
- Enables evaluation on all ranks
- Improves default hyperparams

~45% test acc w/ latest hyperparams on 8 gpus

latest cugraph single node multigpu works but drops degree 0 nodes in neighborloading causing lower test accuracy. @alexbarghi-nv will fix in 24.04 RAPIDS release. in general i think once cugraph 24.02 is out we can merge this PR and the test accuracy will automatically go up when using cugraph once users upgrade to 24.04 cugraph or newer

### CODE REVIEW

File: examples/multi_gpu/papers100m_gcn.py
Comment (akihironitta) [inline examples/multi_gpu/papers100m_gcn.py:18]: It might be worth adding some comment like this:
```suggestion
# Allow computation on objects that are larger than GPU memory
# https://docs.rapids.ai/api/cudf/stable/developer_guide/library_design/#spilling-to-host-memory
os.environ['CUDF_SPILL'] = '1'
```

File: examples/multi_gpu/papers100m_gcn.py
Comment (akihironitta) [inline examples/multi_gpu/papers100m_gcn.py:19]: It'd also be nice to add some comment here as well. Is this for suppressing warnings in the cluster?

File: examples/multi_gpu/papers100m_gcn.py
Comment (akihironitta) [inline examples/multi_gpu/papers100m_gcn.py:261]: small comment: As your current script starts measuring time before launching processes, it should also measure time until after teardown.

**Review by akihironitta — COMMENTED** [review]
This is exciting :) I definitely need some more time to properly review this as I'm not familiar with `dask_cuda`, etc.

Would you be able to benchmark these and share their results (model performance as well as speed improvements) before and after this PR?

File: CHANGELOG.md
Comment (akihironitta) [inline CHANGELOG.md:10]: The changelog entry should be more descriptive IMO. What are the improvements?

**Comment by puririshi98** [comment]
we get up to 5x speed ups with cugraph depending on the system used

File: examples/multi_gpu/papers100m_gcn.py
Comment (puririshi98) [reply examples/multi_gpu/papers100m_gcn.py:21]: @alexbarghi-nv
will provide the answer for this since im not really sure personally. (i also dont know much about dask_cuda😅)

Comment (alexbarghi-nv) [reply examples/multi_gpu/papers100m_gcn.py:21]: CUDF_SPILL enables cudf spilling which saves device memory.  RAPIDS_NO_INITIALIZE ensures that a CUDA context is not created on import of rapids, allowing pytorch to create the context instead.

**Comment by puririshi98** [comment]
@akihironitta ready for review, lmk if anything else needed to merge

**Comment by puririshi98** [comment]
need @alexbarghi-nv to mirror my work on eval/testing on all ranks: https://github.com/pyg-team/pytorch_geometric/pull/8173/commits/c86f852644b3455bdaea6ee8cf8264eed6c5465a
for cugraph
then we will be ready to merge this

**Comment by puririshi98** [comment]
@akihironitta @rusty1s ready to merge

File: CHANGELOG.md
Comment (akihironitta) [inline CHANGELOG.md:10]:
```suggestion
- Added support for cuGraph data loading and `GAT` in single node Papers100m examples ([#8173](https://github.com/pyg-team/pytorch_geometric/pull/8173))
```

File: examples/multi_gpu/papers100m_gcn.py
Comment (akihironitta) [inline examples/multi_gpu/papers100m_gcn.py:24]:
```suggestion
# Ensures that a CUDA context is not created on import of rapids to
# let PyTorch create one instead
# https://docs.rapids.ai/api/dask-cuda/stable/ucx/#software-requirements
os.environ['RAPIDS_NO_INITIALIZE'] = '1'
```

File: examples/ogbn_papers_100m.py
Comment (akihironitta) [inline examples/ogbn_papers_100m.py:127]: Looks like this is not average time but per-iter time. You might need to rerun your experiments if your results are based on this time counter.

**Review by akihironitta — COMMENTED** [review]
LGTM after skimming through the change, but I'm not confident enough to approve this, so I'll leave this to @rusty1s (or future me).

Also, I still feel like it'd be great to have exact numbers (not just approx. improvments in %) you got from your experiments in the PR description.

**Comment by akihironitta** [comment]
@puririshi98 I've just written the PR description, PTAL :)

**Review by alexbarghi-nv — CHANGES_REQUESTED** [review]
Change request; as discussed @puririshi98

File: examples/ogbn_papers_100m.py
Comment (puririshi98) [reply examples/ogbn_papers_100m.py:127]: im a bit confused. the timer starts once we have finished warming up, then at the end of the epoch we check the time elapsed and divide by the number of iters passed since the timer starts. this gives us an average time for each iteration which is what we report

**Comment by puririshi98** [comment]
needs a fix for cugraph from @alexbarghi-nv

**Comment by puririshi98** [comment]
okay @akihironitta @rusty1s, the current code in this PR only works with in 24.02 or newer cugraph so i will circle back for merging this once that is released publicly

**Comment by puririshi98** [comment]
singlenodemultigpu cugraph on latest stack:  
```
Test Accuracy: 12.8855%
```
latest cugraph single node multigpu works and provides a 5.7x speed up in end 2 end run time.
However, the test accuracy  is far too low. needs fix from @alexbarghi-nv

File: examples/README.md
Comment (rusty1s) [inline examples/README.md:15]:
```suggestion
- [`ogbn_papers_100m_cugraph.py`](./ogbn_papers_100m_cugraph.py) shows how to accelerate the `ogbn-papers100m` workflow using [CuGraph](...).
```

File: examples/multi_gpu/README.md
Comment (rusty1s) [inline examples/multi_gpu/README.md:12]:
```suggestion
| [`papers100m_gcn_cugraph.py`](./papers100m_gcn_cugraph.py%60)                      | single-node | Example for accelerating GNN training on `ogbn-papers100M` using [CuGraph](...).                                                                                                                                                                                                                              |
```

File: examples/multi_gpu/papers100m_gcn.py
Comment (rusty1s) [inline examples/multi_gpu/papers100m_gcn.py:27]: What's the change here? I prefer `get_num_workers` over `pyg_num_work`.

File: examples/multi_gpu/papers100m_gcn.py
Comment (rusty1s) [inline examples/multi_gpu/papers100m_gcn.py:64]:
```suggestion
    val_loader = NeighborLoader(data, input_nodes=split_idx['valid'],
```

File: examples/multi_gpu/papers100m_gcn.py
Comment (rusty1s) [inline examples/multi_gpu/papers100m_gcn.py:115]: You need to do `acc.reset()` afterwards.

File: examples/multi_gpu/papers100m_gcn.py
Comment (rusty1s) [inline examples/multi_gpu/papers100m_gcn.py:183]:
```suggestion
                                              heads=args.n_gat_conv_heads,)
    else:
        model = torch_geometric.nn.models.GCN(dataset.num_features,
                                              args.hidden_channels,
                                              args.num_layers,
                                              dataset.num_classes, )
```

File: examples/multi_gpu/papers100m_gcn_cugraph.py
Comment (rusty1s) [inline examples/multi_gpu/papers100m_gcn_cugraph.py:113]: Might be good to leave some more comments here to a new user :)

File: examples/multi_gpu/papers100m_gcn_cugraph.py
Comment (rusty1s) [inline examples/multi_gpu/papers100m_gcn_cugraph.py:139]: Is that necessary?

File: examples/multi_gpu/papers100m_gcn_cugraph.py
Comment (rusty1s) [inline examples/multi_gpu/papers100m_gcn_cugraph.py:189]: Drop?

File: examples/multi_gpu/papers100m_gcn_cugraph.py
Comment (rusty1s) [inline examples/multi_gpu/papers100m_gcn_cugraph.py:226]: Why do we need to call `module`?

File: examples/multi_gpu/papers100m_gcn_cugraph.py
Comment (rusty1s) [inline examples/multi_gpu/papers100m_gcn_cugraph.py:222]: Drop?

File: examples/ogbn_papers_100m.py
Comment (rusty1s) [inline examples/ogbn_papers_100m.py:70]:
```suggestion
                                          dataset.num_classes, ).to(device)
```

File: examples/ogbn_papers_100m.py
Comment (rusty1s) [inline examples/ogbn_papers_100m.py:82]: Drop?

File: examples/ogbn_papers_100m_cugraph.py
Comment (rusty1s) [inline examples/ogbn_papers_100m_cugraph.py:101]: Drop.

File: examples/multi_gpu/papers100m_gcn.py
Comment (puririshi98) [reply examples/multi_gpu/papers100m_gcn.py:108]: this was suggested by @akihironitta when reviewing my mag240m example
"I think `torchmetrics` should handle the reduction by `.compute()`, so there's no need to perform it manually here."

File: examples/multi_gpu/papers100m_gcn_cugraph.py
Comment (puririshi98) [reply examples/multi_gpu/papers100m_gcn_cugraph.py:224]: from cugraph team:
"This is because we have a different number of inputs on each rank. Without it the script will hang forever."
this will be fixed in future release of cugraph.
this is the known issue ticket for cugraph: https://github.com/rapidsai/cugraph/issues/4201

once that is fixed, i can follow up in the future and clean up this part with a quick PR

File: examples/multi_gpu/papers100m_gcn_cugraph.py
Comment (puririshi98) [reply examples/multi_gpu/papers100m_gcn_cugraph.py:144]: currently yes. this is due to limitations with Dask. CuGraph team is trying to remove dependencies on dask. in the future this wont be an issue. we will update example at that time.

**Comment by puririshi98** [comment]
@rusty1s lmk if anything needed to merge

**Comment by rusty1s** [comment]
On it.

**Review by rusty1s — APPROVED** [review]

**Pull request merged**
